orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton.go

```diff
@@ -0,0 +1,185 @@
+	"magma/orc8r/lib/go/util"
+)
+
+// Major TODOs:
+// TODO(reginawang3495): Add save jobs/indexes # times run
+// TODO(reginawang3495): Remove "Desired" from Indexer.Versions once reindexer_queue and queue are removed
+
+// This Reindexer runs as though it is a singleton
+type reindexerSingleton struct {
+	versioner Versioner
+	store     Store
+	// TODO(reginawang3495) Add some structure to save jobs/indexes # times run
+}
+
+// TODO: to be setup and used in M2 Part C
+const reindexLoopInterval = 5 * time.Second
+
+func NewReindexerSingleton(store Store, versioner Versioner) Reindexer {
+	return &reindexerSingleton{store: store, versioner: versioner}
+}
+
+func (r *reindexerSingleton) Run(ctx context.Context) {
+	const indexerID = ""
+	var sendUpdate func(string) = nil
+
+	batches := r.getReindexBatches(ctx)
+	for {
+		if isCanceled(ctx) {
+			glog.Warning("State reindexing async job canceled")
+			return
+		}
+
+		err := r.reindexJobs(ctx, indexerID, batches, sendUpdate)
+
+		// NOTE: consider sleeping when no err for perf reasons
+		if err != nil {
+			clock.Sleep(failedJobSleep)
+		} else {
+			clock.Sleep(reindexLoopInterval)
+		}
+	}
+}
+
+func (r *reindexerSingleton) RunUnsafe(ctx context.Context, indexerID string, sendUpdate func(string)) error {
+	batches := r.getReindexBatches(ctx)
+	return r.reindexJobs(ctx, indexerID, batches, sendUpdate)
+}
+
+func (r *reindexerSingleton) GetIndexerVersions() ([]*indexer.Versions, error) {
+	return r.versioner.GetIndexerVersions()
+}
+
+// Get network-segregated reindex batches with capped number of state IDs per batch.
+func (r *reindexerSingleton) getReindexBatches(ctx context.Context) []reindexBatch {
+	var idsByNetwork state_types.IDsByNetwork
+	for {
+		if isCanceled(ctx) {
+			return nil
+		}
+		ids, err := r.store.GetAllIDs()
+		if err != nil {
+			err = wrap(err, ErrDefault, "")
+			glog.Errorf("Failed to get all state IDs for state indexer reindexing, will retry: %v", err)
+			clock.Sleep(failedGetBatchesSleep)
+			continue
+		}
+		idsByNetwork = ids
+		break
+	}
+
+	var current, rest state_types.IDs
+	var batches []reindexBatch
+	for networkID, ids := range idsByNetwork {
+		rest = ids
+		for len(rest) > 0 {
+			count := util.MinInt(numStatesToReindexPerCall, len(rest))
+			current, rest = rest[:count], rest[count:]
+			batches = append(batches, reindexBatch{networkID: networkID, stateIDs: current})
+		}
+	}
+	return batches
+}
+
+func (r *reindexerSingleton) reindexJobs(ctx context.Context, indexerID string, batches []reindexBatch, sendUpdate func(string)) error {
+	jobs, err := r.getJobs(indexerID)
+	if err != nil {
+		return wrap(err, ErrDefault, indexerID)
+	}
+	// QUESTION: why return error if no jobs? this is brought over from Run / RunUnsafe of reindexer_queue
+	if len(jobs) == 0 {
+		return merrors.ErrNotFound
+	}
+
+	for _, j := range jobs {
+		err = r.reindexJob(j, ctx, batches, sendUpdate)
+		if err != nil {
+			return wrap(err, ErrDefault, indexerID)
```

Comment:
Let's use a multierror here instead of an early return. That is, just because one reindex job fails, that doesn't mean we should give up on the others